octopusdeploy_framework/resource_tenant.go

```diff
@@ -0,0 +1,170 @@
+package octopusdeploy_framework
+
+import (
+	"context"
+	"fmt"
+	"github.com/OctopusDeploy/go-octopusdeploy/v2/pkg/tenants"
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/internal"
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/schemas"
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/util"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"sort"
+)
+
+type tenantTypeResource struct {
+	*Config
+}
+
+func NewTenantResource() resource.Resource {
+	return &tenantTypeResource{}
+}
+
+func (r *tenantTypeResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = util.GetTypeName("tenant")
+}
+
+func (r *tenantTypeResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: schemas.GetTenantResourceSchema(),
+	}
+}
+
+func (r *tenantTypeResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	r.Config = ResourceConfiguration(req, resp)
+}
+
+func (r *tenantTypeResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	internal.Mutex.Lock()
+	defer internal.Mutex.Unlock()
+
+	var data *schemas.TenantModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tenant, err := mapStateToTenant(data)
+	if err != nil {
+		return
+	}
+
+	tflog.Info(ctx, fmt.Sprintf("creating Tenant: %s", tenant.Name))
+
+	createdTenant, err := tenants.Add(r.Config.Client, tenant)
+	if err != nil {
+		resp.Diagnostics.AddError("unable to create tenant", err.Error())
+		return
+	}
+
+	mapTenantToState(data, createdTenant)
+
+	tflog.Info(ctx, fmt.Sprintf("Tenant created (%s)", data.ID))
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *tenantTypeResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var data *schemas.TenantModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tflog.Info(ctx, fmt.Sprintf("reading Tenant (%s)", data.ID))
+
+	client := r.Config.Client
+	tenant, err := tenants.GetByID(client, data.SpaceID.ValueString(), data.ID.ValueString())
+	if err != nil {
```

Comment:
You'll run into the same thing I just ran into with the runbook resource, if a tenant is deleted from Octopus next time you run plan/apply you'll get an error as Terraform can't find the resource. I realised that `errors.ProcessApiErrors` removes the resource from state if the response from the API is a 404.